[torchelastic] Add warning message if `last_call_timeout` for `etcd` rendezvous is less than `2*monitor_interval`

aivanou opened the pull request:
Summary: Add warning message if `last_call_timeout` for `etcd` rendezvous is less than `2*monitor_interval`

Test Plan: unittest

Differential Revision: D33904180

cbalioglu commented on torch/distributed/launcher/api.py:
```diff
@@ -172,6 +172,18 @@ def _get_addr_and_port(
+    monitor_interval = config.monitor_interval
+    if last_call_timeout < monitor_interval * 2:
+        logger.warning(
+            "`last_call_timeout` should be at least `monitor_interval * 2`,"
```
Do you mind briefly explaining why this happens?

aivanou committed:
[torchelastic] Add warning message if `last_call_timeout` for `etcd` rendezvous is less than `2*monitor_interval` (#72098)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/72098

Add warning message if `last_call_timeout` for `etcd` rendezvous is less than `2*monitor_interval`

If the `last_call_timeout` is less than `2*monitor_interval` the nodes may be in different rendezvous version. E.g. Node1 failed, and got restarted. At the same time Node2 continues working. If Node2 does not need communication with Node1, Node2 will not restart its workers, unless TE agent will wake up and check for new rendezvous rounds. TE agent wakes up every `monitor_interval` period.

Node1 will initiate a new rendezvous round. If `nnodes=1:2`, the rendezvous eventually will be in final state waiting for `last_call_timeout`. If the Node1 `monitor_interval` is long enough, the TE agent on that node may not wake up during the new rendezvous round. As a result, Node1 will finish rendezvous round and starts work. When Node2 wakes up, it will not detect a new rendezvous round, since the rendezvous finished.

More info: https://github.com/pytorch/pytorch/issues/67616

Test Plan:
unittest

example of message output:

```
(aivanou-dev) ubuntu@ip-172-31-24-178:~/code/elastic$ CUDA_VISIBLE_DEVICES=1 python -m torch.distributed.run --rdzv_id=14 --max_restarts=100  --rdzv_backend=etcd --rdzv_endpoint=127.0.0.1:2379 --rdzv_conf last_call_timeout=1 --nnodes=1:1 --nproc_per_node=1 --monitor_interval 3 sleep.py
WARNING:torch.distributed.launcher.api:
`last_call_timeout` should be at least `monitor_interval * 2`. Set `last_call_timeout` to a value >= {monitor_interval*2} or use rdzv_backend `etcd-v2`. For details see: https://github.com/pytorch/pytorch/issues/67616.

INFO 2022-02-01 18:52:27,292 Etcd machines: ['http://localhost:2379']
INFO 2022-02-01 18:52:27,297 Attempting to join next rendezvous
```

Differential Revision: D33904180

fbshipit-source-id: d613e567c6a7e319dbcf93e6ac442f19ab20837e
